File: DSA/SakshamKashyap_590014183/day39/day39_Question_3.py
class TwoStacks:
    def __init__(self, n):
        self.size = n
        self.arr = [None] * n
        self.top1 = -1
        self.top2 = self.size

    def push1(self, value):
        """Pushes a value onto the first stack."""
        # Check if there is at least one empty space between the stacks
        if self.top1 < self.top2 - 1:
            self.top1 += 1
            self.arr[self.top1] = value
        else:
            print("Stack Overflow")

    def push2(self, value):
        """Pushes a value onto the second stack."""
        # Check if there is at least one empty space between the stacks
        if self.top1 < self.top2 - 1:
            self.top2 -= 1
            self.arr[self.top2] = value
        else:
            print("Stack Overflow")

    def pop1(self):
        """Pops a value from the first stack."""
        if self.top1 >= 0:
            value = self.arr[self.top1]
            self.top1 -= 1
            return value
        else:
            print("Stack Underflow")
            return None

    def pop2(self):
        """Pops a value from the second stack."""
        if self.top2 < self.size:
            value = self.arr[self.top2]
            self.top2 += 1
            return value
        else:
            print("Stack Underflow")
            return None
    
    def get_stack1(self):
        """Helper to view the contents of stack 1."""
        return self.arr[0:self.top1 + 1]

    def get_stack2(self):
        """Helper to view the contents of stack 2."""
        # The stack grows from right to left, so we reverse it for printing
        return self.arr[self.top2:self.size][::-1]

# --- Test cases ---
ts = TwoStacks(5)

ts.push1(1)
ts.push1(2)
ts.push2(9)
ts.push2(8)

print(f"Stack 1: {ts.get_stack1()}")
print(f"Stack 2: {ts.get_stack2()}")

ts.push1(3)
print(f"Stack 1 after push(3): {ts.get_stack1()}")

# This will cause an overflow
ts.push2(7)

# Pop from stack 2
print(f"Popped from Stack 2: {ts.pop2()}")
print(f"Stack 2 after pop: {ts.get_stack2()}")

File: DSA/Rahul_Mankani_590014725/Day39_Question3.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 1000
typedef struct 
{
    int a[MAX], t1, t2;
} TwoStacks;
void init(TwoStacks* ts) 
{
    ts->t1 = -1;
    ts->t2 = MAX;
}
int push1(TwoStacks* ts, int x) 
{
    if(ts->t1 + 1 < ts->t2) 
    {
        ts->a[++ts->t1] = x;
        return 1;
    }
    return 0;
}
int push2(TwoStacks* ts, int x) 
{
    if(ts->t2 - 1 > ts->t1) 
    {
        ts->a[--ts->t2] = x;
        return 1;
    }
    return 0;
}
int pop1(TwoStacks* ts) 
{
    if(ts->t1 >= 0) return ts->a[ts->t1--];
    return -1;
}
int pop2(TwoStacks* ts) 
{
    if(ts->t2 < MAX) return ts->a[ts->t2++];
    return -1;
}
int main() 
{
    TwoStacks ts;
    init(&ts);
    int q, s, x;
    scanf("%d", &q);
    while(q--) 
    {
        scanf("%d%d", &s, &x);
        if(s == 1) 
        {
            printf("%d\n", x == 1 ? push1(&ts, x) : pop1(&ts));
        } 
        else 
        {
            printf("%d\n", x == 1 ? push2(&ts, x) : pop2(&ts));
        }
    }
    return 0;
}
